src/Uno.Foundation/Collections/Internal/NullableKeyDictionary.cs

```diff
@@ -0,0 +1,184 @@
+﻿#nullable enable
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Uno.Foundation.Collections;
+
+/// <summary>
+/// Represents a variant of dictionary that allows adding a value for a null key.
+/// </summary>
+/// <typeparam name="TKey">Key.</typeparam>
+/// <typeparam name="TValue">Value.</typeparam>
+internal class NullableKeyDictionary<TKey, TValue> : IDictionary<TKey, TValue>
+	  where TKey : class
+{
+	private readonly Dictionary<TKey, TValue> _dictionary;
+	private TValue _nullKeyValue = default!;
+	private bool _containsNullKey;
+
+	public NullableKeyDictionary() =>
+		_dictionary = new Dictionary<TKey, TValue>();
+
+	public NullableKeyDictionary(IEqualityComparer<TKey> comparer) =>
+		_dictionary = new Dictionary<TKey, TValue>(comparer);
+
+	public bool ContainsKey(TKey? key) => key == null ? _containsNullKey : _dictionary.ContainsKey(key);
+
+	public void Add(TKey? key, TValue value)
+	{
+		if (key == null)
+		{
+			if (_containsNullKey)
+			{
+				throw new ArgumentException("Value already present");
+			}
+
+			_nullKeyValue = value;
+			_containsNullKey = true;
+		}
+		else
+		{
+			_dictionary.Add(key, value);
+		}
+	}
+
+	public bool Remove(TKey? key)
+	{
+		if (key != null)
+		{
+			return _dictionary.Remove(key);
+		}
+
+		if (_containsNullKey)
+		{
+			_nullKeyValue = default!;
+			_containsNullKey = false;
+			return true;
+		}
+
+		return false;
+	}
+
+	public bool TryGetValue(TKey? key, out TValue value)
+	{
+		if (key != null)
+		{
+			return _dictionary.TryGetValue(key, out value!);
+		}
+		else
+		{
+			value = _containsNullKey ? _nullKeyValue : default!;
```

Comment:
I guess that should work?

```suggestion
			value = _nullKeyValue;
```